src/lib/client.js

```diff
@@ -26,18 +30,93 @@ module.exports = class BtpClient {
     this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
   }
 
+  tryToOpenWebSocketClient () {
+    return new Promise((resolve, reject) => {
+      const ws = new WebSocket(this._wsUri, {
+        perMessageDeflate: false
+      })
+      console.log('try to open', this._wsUri)
+      let hasBeenOpen = false
+      ws.on('open', () => {
+        console.log('open!')
+        hasBeenOpen = true
+        resolve(ws)
+      })
+      ws.on('error', (err) => {
+        if (hasBeenOpen) {
+          console.error('WebSocket error', err)
+        } else {
+          reject(err)
+        }
+      })
+      ws.on('close', () => {
+        console.log('close!')
+        // automatically reconnect if server reboots
+        if (hasBeenOpen && !ws.shouldClose) {
+          this.ensureUpstream()
+        }
+      })
+    })
+  }
+
+  getOpenWebSocketClient () {
+    return new Promise((resolve) => {
+      let done = false
+      const tryOnce = () => {
+        this.tryToOpenWebSocketClient().then(ws => {
+          if (done) { // this can happen if opening the WebSocket works, but just takes long
+            console.log('got it! but was already done')
+            ws.shouldClose = true
+            ws.close()
+          } else {
+            console.log('got it! setting done')
+            done = true
+            clearInterval(timer)
+            resolve(ws)
+          }
+        }).catch((err) => {
+          console.error('error caught while trying to open WebSocket client', err)
+        })
+      }
+      let timer = setInterval(tryOnce, CONNECT_RETRY_INTERVAL)
```

Comment:
it looks like the default `CONNECT_RETRY_INTERVAL` is 1000ms. What happens if the websocket takes more than 1000ms to open, but does open successfully? There would be overlapping connection attempts, but I'm not sure what the result would be.

Reply:
that's handled by https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/blob/a24a4d7a524296bc67f859cf52e5d09b61048b98/src/lib/client.js#L67